src/shell/shell.cpp

```diff
@@ -178,141 +178,269 @@ DOS_Shell::DOS_Shell()
           call(false)
 {}
 
-// TODO: this function should be refactored to make to it easier to understand.
-// It's currently riddled with pointer and array adjustments each loop plus
-// branches and sub-loops.
-Bitu DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, bool *append)
+void DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, char **pipe, bool *append)
 {
-	char * lr=s;
-	char * lw=s;
+	char *lr = s;
+	char *lw = s;
 	char ch;
-	Bitu num=0;
 	bool quote = false;
 	char *temp = nullptr;
-	size_t temp_len = 0;
-
-	while ( (ch=*lr++) ) {
-		if(quote && ch != '"') { /* don't parse redirection within quotes. Not perfect yet. Escaped quotes will mess the count up */
+	char **redir = nullptr;
+	size_t found, temp_len;
+	std::string str, chrs;
+
+	while ((ch = *lr++)) {
+		if (quote && ch != '"') { /* don't parse redirection within
+			                     quotes. Not perfect yet. Escaped
+			                     quotes will mess the count up */
 			*lw++ = ch;
 			continue;
 		}
-
-		switch (ch) {
-		case '"':
+		if (ch == '"') {
 			quote = !quote;
-			break;
-		case '>':
-			*append=((*lr)=='>');
-			if (*append) lr++;
-			lr=ltrim(lr);
-			if (*ofn) {
-				delete[] * ofn;
-				*ofn = nullptr;
+		} else if (ch == '>' || ch == '<' || ch == '|') {
+			// Overwrite with >, and append with >>
+			if (ch == '>' && (*append = (*lr == '>')))
+				lr++;
+			redir = ch == '>' ? ofn : (ch == '<' ? ifn : pipe);
+			if (*redir) {
+				delete[] * redir;
+				*redir = nullptr;
 			}
-			*ofn = lr;
-			while (*lr && *lr!=' ' && *lr!='<' && *lr!='|') lr++;
-			//if it ends on a : => remove it.
-			if((*ofn != lr) && (lr[-1] == ':')) lr[-1] = 0;
-			temp_len = static_cast<size_t>(lr - *ofn + 1u);
-			temp = new char[temp_len];
-			safe_strncpy(temp, *ofn, temp_len);
-			*ofn = temp;
+			lr = ltrim(lr);
+			str = *redir = lr;
+			chrs = ch == '|' ? "" : (ch != '<' ? " |<" : " |>");
+			found = str.find_first_of(chrs);
+			if (found == std::string::npos)
+				temp_len = str.size();
+			else
+				temp_len = found - (str[found - 1] == ':' ? 1 : 0);
+			lr += temp_len;
+			str[temp_len] = 0;
+			temp = new char[temp_len + 1];
+			safe_strncpy(temp, str.c_str(), temp_len + 1);
+			*redir = temp;
 			continue;
+		}
+		*lw++ = ch;
+	}
+	*lw = 0;
+}
 
-		case '<':
-			if (*ifn) {
-				delete[] * ifn;
-				*ifn = nullptr;
+bool get_pipe_status(const char *out,
+                     const char *pipe,
+                     char *pipetmp,
+                     const bool append,
+                     bool &failed_pipe)
+{
+	uint16_t fattr = 0;
+	uint16_t dummy = 0;
+	uint16_t dummy2 = 0;
+	uint32_t bigdummy = 0;
```

Comment:
Even if these are unused, maybe the actual names could be used - unless they're used in lots of places with different meanings?
I'm seeing the dummies used in opening, seeking, and open-extended: I guess these variables get some status written into them that we don't care about?

Reply:
Indeed, these variables are used by various file-related functions and there is no need to actually care about their values. This is why they are named so.